src/service/url-replacements-impl.js

```diff
@@ -408,55 +489,158 @@ export class UrlReplacements {
 
     if (isNaN(metric) || metric == Infinity) {
       // The metric is not supported.
-      return Promise.resolve();
+      return;
     } else if (metric < 0) {
-      // Metric is not yet available. Retry after a delay.
-      return loadPromise(this.win_).then(() => {
-        metric = (endEvent === undefined)
-            ? timingInfo[startEvent]
-            : timingInfo[endEvent] - timingInfo[startEvent];
-        return (isNaN(metric) || metric == Infinity || metric < 0)
-            ? undefined
-            : String(metric);
-      });
+      return '';
     } else {
-      return Promise.resolve(String(metric));
+      return String(metric);
     }
   }
 
   /**
    * Returns navigation information from the current browsing context.
    * @param {string} attribute
-   * @return {!Promise<string|undefined>}
+   * @return {string|undefined}
    * @private
    */
   getNavigationData_(attribute) {
     const navigationInfo = this.win_['performance']
         && this.win_['performance']['navigation'];
     if (!navigationInfo || navigationInfo[attribute] === undefined) {
       // PerformanceNavigation interface is not supported or attribute is not implemented.
-      return Promise.resolve();
+      return;
     }
 
     return String(navigationInfo[attribute]);
   }
 
   /**
-   * Sets the value resolver for the variable with the specified name. The
-   * value resolver may optionally take an extra parameter.
+   * Sets a synchronous value resolver for the variable with the specified name.
+   * The value resolver may optionally take an extra parameter.
+   * Can be called in conjuction with setAsync to allow for additional
+   * asynchronous resolver where expand will use async and expandSync the sync
+   * version.
    * @param {string} varName
-   * @param {function(*):*} resolver
+   * @param {!SyncResolverDef} syncResolver
    * @return {!UrlReplacements}
    * @private
    */
-  set_(varName, resolver) {
+  set_(varName, syncResolver) {
     dev.assert(varName.indexOf('RETURN') == -1);
-    this.replacements_[varName] = resolver;
+    this.replacements_[varName] = this.replacements_[varName] || {};
```

Comment:
It'd be better to use a uniform object `{sync: undefined, async: undefined}`, that way we don't modify its hidden class.